ipn/ipnlocal/local.go

```diff
@@ -6203,3 +6204,157 @@ func mayDeref[T any](p *T) (v T) {
+					distance := longLatDistance(derpHomeCoordinates, candidateCoordinates)
+					candidateDistanceMap[distance] = append(candidateDistanceMap[distance], candidate)
+					distances = append(distances, distance)
+				}
+			}
+		}
+	}
+	sortedRegions = sortRegions(sortedRegions, lastReport)
+	if len(sortedRegions) > 0 && candidateLatencyMap[sortedRegions[0]] != nil && candidateLatencyMap[sortedRegions[0]][0].Valid() {
+		preferredExitNodeID = candidateLatencyMap[sortedRegions[0]][0].StableID()
+		preferredExitNodeName = candidateLatencyMap[sortedRegions[0]][0].Name()
+		hi := candidateLatencyMap[sortedRegions[0]][0].Hostinfo()
+		if hi.Valid() && hi.Location().View().Valid() {
+			preferredExitNodeLocation = *hi.Location()
+		}
+	} else {
+		slices.Sort(distances)
+		if len(distances) > 0 {
+			if len(candidateDistanceMap[distances[0]]) == 1 && candidateDistanceMap[distances[0]][0].Valid() {
+				preferredExitNodeID = candidateDistanceMap[distances[0]][0].StableID()
+				preferredExitNodeName = candidateDistanceMap[distances[0]][0].Name()
+				preferredExitNodeLocation = *candidateDistanceMap[distances[0]][0].Hostinfo().Location()
+			} else {
+				chosen := pickWeighted(candidateDistanceMap[distances[0]])
+				if chosen.Valid() {
+					preferredExitNodeID = chosen.StableID()
+					preferredExitNodeName = chosen.Name()
+					preferredExitNodeLocation = *chosen.Hostinfo().Location()
+				}
+			}
+		}
+	}
+	if preferredExitNodeID.IsZero() {
+		return preferredExitNodeID, "", preferredExitNodeLocation, errNoSuggestion
+	}
+	return preferredExitNodeID, preferredExitNodeName, preferredExitNodeLocation, nil
+}
+
+func pickWeighted(candidates []tailcfg.NodeView) tailcfg.NodeView {
+	maxWeight := 0
+	var chosenCandidate tailcfg.NodeView
+	for _, candidate := range candidates {
+		if candidate.Hostinfo().Location().View().Valid() && candidate.Hostinfo().Location().Priority > maxWeight {
+			maxWeight = candidate.Hostinfo().Location().Priority
+			chosenCandidate = candidate
+		}
+	}
+	return chosenCandidate
+}
+
+// sortRegions returns a list of sorted regions by ascending latency given a list of region IDs and a netcheck report.
+func sortRegions(regions []int, lastReport *netcheck.Report) []int {
+	sort.Slice(regions, func(i, j int) bool {
+		iLatency, ok := lastReport.RegionLatency[regions[i]]
+		if !ok || iLatency == 0 {
+			iLatency = math.MaxInt
+		}
+		jLatency, ok := lastReport.RegionLatency[regions[j]]
+		if !ok || jLatency == 0 {
+			jLatency = math.MaxInt
+		}
+		return iLatency < jLatency
+	})
+	return regions
+}
+
+// longLatDistance returns an estimated distance given two lists of float64 values that represent geographic coordinates.
+// The first index is latitude and the second index is longitude.
+// Value is returned in meters. We tolerate up to 100km difference in accuracy.
+func longLatDistance(firstDistance []float64, secondDistance []float64) float64 {
```

Comment:
`firstDistance` and `secondDistance` are not distances. They are coordinates in degrees.

`from, to [2]float64` maybe, or even `fromLat, fromLong, toLat, toLong float64`